base/file.jl

```diff
@@ -913,7 +916,54 @@ julia> readdir(abspath("base"), join=true)
  "/home/JuliaUser/dev/julia/base/weakkeydict.jl"
 ```
 """
-function readdir(dir::AbstractString; join::Bool=false, sort::Bool=true)
+readdir(; join::Bool=false, kwargs...) = readdir(join ? pwd() : "."; join, kwargs...)::Vector{String}
+readdir(dir::AbstractString; kwargs...) = _readdir(dir; return_objects=false, kwargs...)::Vector{String}
+
+# this might be better as an Enum but they're not available here
+# UV_DIRENT_T
+const UV_DIRENT_UNKNOWN = Cint(0)
+const UV_DIRENT_FILE = Cint(1)
+const UV_DIRENT_DIR = Cint(2)
+const UV_DIRENT_LINK = Cint(3)
+const UV_DIRENT_FIFO = Cint(4)
+const UV_DIRENT_SOCKET = Cint(5)
+const UV_DIRENT_CHAR = Cint(6)
+const UV_DIRENT_BLOCK = Cint(7)
+struct FileKind
+    dir::String
+    name::String
+    rawtype::Cint
+end
+function Base.getproperty(obj::FileKind, p::Symbol)
+    if p === :path
+        return joinpath(obj.dir, obj.name)
+    else
+        return getfield(obj, p)
+    end
+end
+Base.isless(a::FileKind, b::FileKind) = isless(a.name, b.name)
+Base.hash(o::FileKind, h::UInt) = hash(o.dir, hash(o.name, hash(o.rawtype, h)))
+==(a::FileKind, b::FileKind) = a.name == b.name && a.dir == b.dir && a.rawtype == b.rawtype
+joinpath(obj::FileKind, args...) = joinpath(obj.path, args...)
+isunknown(obj::FileKind) =  obj.rawtype == UV_DIRENT_UNKNOWN
+islink(obj::FileKind) =     isunknown(obj) ? islink(obj.path) : obj.rawtype == UV_DIRENT_LINK
+isfile(obj::FileKind) =     (isunknown(obj) || islink(obj)) ? isfile(obj.path)      : obj.rawtype == UV_DIRENT_FILE
+isdir(obj::FileKind) =      (isunknown(obj) || islink(obj)) ? isdir(obj.path)       : obj.rawtype == UV_DIRENT_DIR
+isfifo(obj::FileKind) =     (isunknown(obj) || islink(obj)) ? isfifo(obj.path)      : obj.rawtype == UV_DIRENT_FIFO
+issocket(obj::FileKind) =   (isunknown(obj) || islink(obj)) ? issocket(obj.path)    : obj.rawtype == UV_DIRENT_SOCKET
+ischardev(obj::FileKind) =  (isunknown(obj) || islink(obj)) ? ischardev(obj.path)   : obj.rawtype == UV_DIRENT_CHAR
+isblockdev(obj::FileKind) = (isunknown(obj) || islink(obj)) ? isblockdev(obj.path)  : obj.rawtype == UV_DIRENT_BLOCK
+
+"""
+    readdirx(dir::AbstractString=pwd();
+        join::Bool = false,
+        sort::Bool = true,
+    ) -> Vector{FileKind}
+"""
+readdirx(; join::Bool=false, kwargs...) = readdirx(join ? pwd() : "."; join, kwargs...)::Vector{FileKind}
+readdirx(dir::AbstractString; kwargs...) = _readdir(dir; return_objects=true, kwargs...)::Vector{FileKind}
+
+function _readdir(dir::AbstractString; return_objects::Bool=false, join::Bool=false, sort::Bool=true)
```

Comment:
Yeah. It changes whether dir is `"."` or the absolute `pwd()`, even if it doesn't have other effects because the struct keeps dir and name separate. `readdirx` could just always use `pwd()` perhaps, and not have a join kwarg?